poppy/wavefront.py

```diff
@@ -0,0 +1,359 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import scipy.constants as const
+import astropy.units as u
+from copy import deepcopy
+from scipy.linalg import lstsq
+
+from poppy.fresnel import FresnelWavefront, QuadraticLens
+
+
+class PhysicalFresnelWavefront(FresnelWavefront):
+    """
+    This class extends the capabilities of poppy's FresnelWavefront class.
+    
+    The extension includes the redefinition of the wavefront array to
+    represent an electric field with the appropriate units V.m^-1.
+    Some utility functions have been defined to be able to compute the beam
+    radius and beam quality factor. Additionally, a wavelength scaling has 
+    been introduced to simulate a beam quality factor.
+    
+    Parameters
+    ----------
+    M2 : float
+        Beam quality factor.
+    n0 : float
+        Ambient refractive index.
+    """
+    
+    def __init__(self,
+                 beam_radius,
+                 units=u.m,
+                 rayleigh_factor=2.0,
+                 oversample=2,
+                 wavelength=1064.0e-9,
+                 npix=1024,
+                 M2=1.0,
+                 n0=1.00027398, # refractive index of air @ 15°C, lambda=1064nm
+                 **kwargs):
+        super(PhysicalFresnelWavefront, self).__init__(
+                beam_radius=beam_radius,
+                units=units,
+                rayleigh_factor=rayleigh_factor,
+                oversample=oversample,
+                wavelength=wavelength*M2/n0, # scale the wavelength and apply refractive index
+                npix=npix,
+                **kwargs
+                )
+        self.lam = wavelength/n0 # its unscaled value is needed to compute M2
+        self.n0 = n0
+        self.npix = npix*oversample
+        
+        return
+    
+    @property
+    def x(self):
+        """Spatial grid (m)."""
+        y, x = self.coordinates()
+        
+        return x[0, :]
+    
+    @property
+    def dx(self):
+        """Spatial grid sampling (m)."""
+        
+        return self.pixelscale.to(u.m/u.pixel).value
+    
+    @property
+    def xy(self):
+        """Spatial grid (m) as 2 dimensional array."""
+        y, x = self.coordinates()
+        
+        return np.array([x.flatten(), y.flatten()]).T
+        
+    @property
+    def q(self):
+        """Momentum grid (m^-1)."""
+        
+        return np.fft.fftfreq(self.npix, d=self.dx)*2.0*np.pi
+    
+    @property
+    def dq(self):
+        """Momentum grid sampling (m^-1)."""
+        
+        return 2.0*np.pi/self.npix/self.dx
+    
+    @property
+    def intensity(self):
+        """Intensity distribution (W.m^-2)."""
+        
+        return const.c * self.n0 * const.epsilon_0 * np.abs(self.amplitude)**2 / 2.0
+    
+    @property
+    def power(self):
+        """Power of the wavefront (W)."""
+        
+        intensity = self.intensity
+        dx = self.dx
+        
+        return dx**2*np.sum(intensity)
+    
+    def scale(self, P):
```

Comment:
`scale` is a very generic name for this function and doesn't make clear what is being scaled. Perhaps `scale_power` or `set_power`? 

Relatedly there's the `normalize` function in the regular Wavefront class. That's not relevant for a `PhysicalFresnelWavefront` so you could consider adding code to make that explicit, and ignore any attempt to do that kind of normalization on this class:  
```
   def normalize(self):
        # for PhysicalFresnelWavefronts use scale_power instead of 
        # normalizing total intensity to 1
        pass
```